kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -117,307 +135,138 @@ public suspend inline fun <T> Mutex.withLock(owner: Any? = null, action: () -> T
-                            state.owner = waiter.owner ?: LOCKED
-                            waiter.completeResumeLockWaiter()
-                            return
-                        }
-                    }
-                }
-                else -> error("Illegal state $state")
-            }
+        while (true) {
+            // Is this mutex locked?
+            check(isLocked) { "This mutex is not locked" }
+            // Read the owner, waiting until it is set in a spin-loop if required.
+            val curOwner = this.owner.value
+            if (curOwner === NO_OWNER) continue // <-- ATTENTION, BLOCKING PART HERE
+            // Check the owner.
+            check(curOwner === owner) { "This mutex is locked by $curOwner, but $owner is expected" }
+            // Try to clean the owner first. We need to use CAS here to synchronize with concurrent `unlock(..)`-s.
+            if (!this.owner.compareAndSet(owner, NO_OWNER)) continue
+            // Release the semaphore permit at the end.
+            release()
+            return
         }
     }
 
-    override fun toString(): String {
-        _state.loop { state ->
-            when (state) {
-                is Empty -> return "Mutex[${state.locked}]"
-                is OpDescriptor -> state.perform(this)
-                is LockedQueue -> return "Mutex[${state.owner}]"
-                else -> error("Illegal state $state")
-            }
-        }
-    }
+    @Suppress("UNCHECKED_CAST", "OverridingDeprecatedMember")
+    override val onLock: SelectClause2<Any?, Mutex> get() = SelectClause2Impl(
+        clauseObject = this,
+        regFunc = MutexImpl::onLockRegFunction as RegistrationFunction,
+        processResFunc = MutexImpl::onLockProcessResult as ProcessResultFunction,
+        onCancellationConstructor = onCancellationUnlockConstructor
+    )
 
-    private class LockedQueue(
-        @JvmField var owner: Any
-    ) : LockFreeLinkedListHead() {
-        override fun toString(): String = "LockedQueue[$owner]"
+    protected open fun onLockRegFunction(select: SelectInstance<*>, owner: Any?) {
+        onAcquireRegFunction(SelectInstanceWithOwner(select, owner), owner)
     }
 
-    private abstract inner class LockWaiter(
-        @JvmField val owner: Any?
-    ) : LockFreeLinkedListNode(), DisposableHandle {
-        private val isTaken = atomic(false)
-        fun take(): Boolean = isTaken.compareAndSet(false, true)
-        final override fun dispose() { remove() }
-        abstract fun tryResumeLockWaiter(): Boolean
-        abstract fun completeResumeLockWaiter()
+    protected open fun onLockProcessResult(owner: Any?, result: Any?): Any? {
+        return this
     }
 
-    private inner class LockCont(
-        owner: Any?,
-        private val cont: CancellableContinuation<Unit>
-    ) : LockWaiter(owner) {
+    private val onCancellationUnlockConstructor: OnCancellationConstructor =
```

Comment:
It is easier to read the code when a logical part is not split into pieces. 

The pattern is to allocate `OnCancellationConstructor` at once, reusing it on each `onLock` invocation. I would prefer to keep the code consistent in all implementations instead of trying to save a couple of bytes here.